### PR TITLE
Fixed bug in parallelization in MPS simulator. 

### PR DESCRIPTION
**Pull request by merav-aharoni**
### Summary
There was an "omp parallel" on a for loop that had a vector that was assigned values using push_back. That caused undeterministic behavior. Changed the push_back to assignment.

### CODE REVIEW

File: src/simulators/tensor_network/matrix_product_state.cpp [hunk 1]
```diff
@@ -92,6 +92,12 @@ cmatrix_t mul_matrix_by_lambda(const cmatrix_t &mat,
   if (lambda == rvector_t {1.0}) return mat;
   cmatrix_t res_mat(mat);
   uint_t num_rows = mat.GetRows(), num_cols = mat.GetColumns();
+
+  #ifdef _WIN32
+     #pragma omp parallel for
+  #else
+     #pragma omp parallel for collapse(2)
+  #endif
   for(uint_t row = 0; row < num_rows; row++) {
```
Comment (atilag) [inline src/simulators/tensor_network/matrix_product_state.cpp]: We need to change the type of `row` to `int_t` due to incompatibilities with Windows OpenMP support.

File: src/simulators/tensor_network/matrix_product_state.cpp [hunk 2]
```diff
@@ -92,6 +92,12 @@ cmatrix_t mul_matrix_by_lambda(const cmatrix_t &mat,
...
       for(uint_t col = 0; col < num_cols; col++) {
```
Comment (atilag) [inline src/simulators/tensor_network/matrix_product_state.cpp]: We need to chage this type too: `uint_t col` => `int_t col`

File: src/simulators/tensor_network/matrix_product_state.cpp (hunk 1 above)
Comment (merav-aharoni) [reply src/simulators/tensor_network/matrix_product_state.cpp]: Corrected both places as you pointed out.
It is a problem to catch failures that only happen non-deterministically. I can only suggest running the unittest a few times.
We also use another kind of testing here - we create random circuits and then compare results between the mps and the statevector. If you are interested, I can add these somewhere as well.

File: src/simulators/tensor_network/matrix_product_state.cpp (hunk 2 above)
Comment (merav-aharoni) [reply src/simulators/tensor_network/matrix_product_state.cpp]: see above.

File: CHANGELOG.md [hunk 3]
```diff
@@ -35,6 +35,7 @@ Removed
 
 Fixed
 -----
+- Bug in handling parallelization in matrix_product_state.cpp
```
Comment (atilag) [inline CHANGELOG.md]: We need to add the number of the PR at the end of this line (see the other entries) :)

Comment (merav-aharoni) [reply CHANGELOG.md]: right. done.

**Pull request merged**